File: go/plugins/googlecloud/metrics.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package googlecloud

import (
	"context"
	"fmt"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/metric"
)

// MetricCounterOptions holds configuration for a counter metric
type MetricCounterOptions struct {
	Description string
	Unit        string
}

// MetricHistogramOptions holds configuration for a histogram metric
type MetricHistogramOptions struct {
	Description string
	Unit        string
}

// MetricCounter wraps OpenTelemetry counter with Genkit conventions
type MetricCounter struct {
	counter metric.Int64Counter
}

// NewMetricCounter creates a new counter metric with the given name and options
func NewMetricCounter(name string, opts MetricCounterOptions) *MetricCounter {
	meter := otel.Meter("genkit")
	counter, err := meter.Int64Counter(name,
		metric.WithDescription(opts.Description),
		metric.WithUnit(opts.Unit))
	if err != nil {
		// In production, we might want to handle this differently
		panic(fmt.Sprintf("failed to create counter %s: %v", name, err))
	}
	return &MetricCounter{counter: counter}
}

// Add records a value to the counter with the given attributes
func (m *MetricCounter) Add(value int64, attributes map[string]interface{}) {
	if m.counter == nil {
		return
	}
	attrs := convertToOTelAttributes(attributes)
	m.counter.Add(context.Background(), value, metric.WithAttributes(attrs...))
}

// MetricHistogram wraps OpenTelemetry histogram with Genkit conventions
type MetricHistogram struct {
	histogram metric.Float64Histogram
}

// NewMetricHistogram creates a new histogram metric with the given name and options
func NewMetricHistogram(name string, opts MetricHistogramOptions) *MetricHistogram {
	meter := otel.Meter("genkit")
	histogram, err := meter.Float64Histogram(name,
		metric.WithDescription(opts.Description),
		metric.WithUnit(opts.Unit))
	if err != nil {
		// In production, we might want to handle this differently
		panic(fmt.Sprintf("failed to create histogram %s: %v", name, err))
	}
	return &MetricHistogram{histogram: histogram}
}

// Record records a value to the histogram with the given attributes
func (m *MetricHistogram) Record(value float64, attributes map[string]interface{}) {
	if m.histogram == nil {
		return
	}
	attrs := convertToOTelAttributes(attributes)
	m.histogram.Record(context.Background(), value, metric.WithAttributes(attrs...))
}

// convertToOTelAttributes converts a map of string interfaces to OpenTelemetry attributes
func convertToOTelAttributes(attrs map[string]interface{}) []attribute.KeyValue {
	if attrs == nil {
		return nil
	}

	result := make([]attribute.KeyValue, 0, len(attrs))
	for key, value := range attrs {
		if value == nil {
			continue
		}

		switch v := value.(type) {
		case string:
			result = append(result, attribute.String(key, v))
		case int:
			result = append(result, attribute.Int(key, v))
		case int64:
			result = append(result, attribute.Int64(key, v))
		case float64:
			result = append(result, attribute.Float64(key, v))
		case bool:
			result = append(result, attribute.Bool(key, v))
		default:
			// Convert to string for unsupported types
			result = append(result, attribute.String(key, fmt.Sprintf("%v", v)))
		}
	}
	return result
}

// internalMetricNamespaceWrap prefixes metric names with genkit namespace
func internalMetricNamespaceWrap(namespace, metricName string) string {
	return fmt.Sprintf("genkit/%s/%s", namespace, metricName)
}

File: js/plugins/google-cloud/src/metrics.ts
/**
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  metrics,
  type Counter,
  type Histogram,
  type Meter,
} from '@opentelemetry/api';
import type { ReadableSpan } from '@opentelemetry/sdk-trace-base';

export const METER_NAME = 'genkit';
export const METRIC_NAME_PREFIX = 'genkit';
const METRIC_DIMENSION_MAX_CHARS = 256;

export function internalMetricNamespaceWrap(...namespaces) {
  return [METRIC_NAME_PREFIX, ...namespaces].join('/');
}

type MetricCreateFn<T> = (meter: Meter) => T;

/**
 * Wrapper for OpenTelemetry metrics.
 *
 * The OpenTelemetry {MeterProvider} can only be accessed through the metrics
 * API after the NodeSDK library has been initialized. To prevent race
 * conditions we defer the instantiation of the metric to when it is first
 * ticked.
 */
class Metric<T> {
  readonly createFn: MetricCreateFn<T>;
  readonly meterName: string;
  metric?: T;

  constructor(createFn: MetricCreateFn<T>, meterName: string = METER_NAME) {
    this.meterName = meterName;
    this.createFn = createFn;
  }

  get(): T {
    if (!this.metric) {
      this.metric = this.createFn(
        metrics.getMeterProvider().getMeter(this.meterName)
      );
    }

    return this.metric;
  }
}

/**
 * Wrapper for an OpenTelemetry Counter.
 *
 * By using this wrapper, we defer initialization of the counter until it is
 * need, which ensures that the OpenTelemetry SDK has been initialized before
 * the metric has been defined.
 */
export class MetricCounter extends Metric<Counter> {
  constructor(name: string, options: any) {
    super((meter) => meter.createCounter(name, options));
  }

  add(val?: number, opts?: any) {
    if (val) {
      truncateDimensions(opts);
      this.get().add(val, opts);
    }
  }
}

/**
 * Wrapper for an OpenTelemetry Histogram.
 *
 * By using this wrapper, we defer initialization of the counter until it is
 * need, which ensures that the OpenTelemetry SDK has been initialized before
 * the metric has been defined.
 */
export class MetricHistogram extends Metric<Histogram> {
  constructor(name: string, options: any) {
    super((meter) => meter.createHistogram(name, options));
  }

  record(val?: number, opts?: any) {
    if (val) {
      truncateDimensions(opts);
      this.get().record(val, opts);
    }
  }
}

/**
 * Truncates all of the metric dimensions to ensure they are below the trace
 * attribute max. Labels should be consistent between metrics and traces so that
 * data can be properly correlated. This will truncate all dimensions for
 * metrics such that they will be consistent with data included in traces.
 */
function truncateDimensions(opts?: any) {
  if (opts) {
    Object.keys(opts).forEach((k) => {
      if (typeof opts[k] === 'string') {
        opts[k] = opts[k].substring(0, METRIC_DIMENSION_MAX_CHARS);
      }
    });
  }
}

export interface Telemetry {
  tick(span: ReadableSpan, logIO: boolean, projectId?: string): void;
}
